src/iterator/backend.rs

```diff
@@ -78,74 +71,41 @@ impl Drop for RegisteredSignals {
     }
 }
 
-/// A struct for delivering received signals to the main program flow.
-/// The self-pipe IO type is generic. See the [with_pipe](#method.with_pipe)
-/// method for requirements for the IO type.
+/// A struct to control an associated [`SignalDelivery`] or
+/// [`SignalIterator`] instance.
+///
+/// It allows to register more signal handlers and to shutdown the signal
+/// delivery. You may retrieve an instance of this type from the
+/// [`SignalDelivery::controller`] method.
 #[derive(Debug)]
-pub struct SignalDelivery<R, W> {
-    ids: Arc<RegisteredSignals>,
-    waker: Arc<Waker<R, W>>,
-}
-
-// We have to implement Clone by hand due to a known issue of
-// derive(Clone) for generic structs.
-impl<R, W> Clone for SignalDelivery<R, W> {
-    fn clone(&self) -> Self {
-        Self {
-            ids: Arc::clone(&self.ids),
-            waker: Arc::clone(&self.waker),
-        }
-    }
+pub struct Controller<W> {
+    waker: Arc<Waker<W>>,
+    closed: AtomicBool,
+    registerd_signal_ids: RegisteredSignals,
 }
 
-impl<R, W> SignalDelivery<R, W>
+impl<W> Controller<W>
```

Comment:
> I like the technical part, but I'd suggest few small details around the API.
> 
>     * I'm not sure if `Controller` is the best name. Aren't these things usually called `Handle`?

I think of the type as some kind of remote control for `SignalDelivery` instances. Maybe `DeliveryControl` is a better name? This describes its purpose better.

>     * I'd suggest to make that `Controller`/`Handle` into a newtype (to not show the detail of the `Arc`) and to hide the type parameter. Some kind of `dyn` should solve it, and users are likely not interested in that O:-)

Also see my comment https://github.com/vorner/signal-hook/pull/49#discussion_r507989437. If we wrap the whole thing in a new type we may even get away without a Box<dyn>. I'll do this.

>     * As for the exposure of the dependency ‒ I think each crate can re-export the `Controller`/`Handle` type. That way the user can just import `signal-hook-tokio` or something like that.


My concerns about the API and knowing dependencies of dependencies was more related to promoting method chains like `signals.controller().close()` on the client side.

>     * Would it be possible to still be able to clone the iterator? Most of the reader types have some kind of `try_clone`.

Because the `SignalDelivery` struct is generic we need a trait to do the cloning there. But the `try_clone` method isn't part of any trait. So we would have to define our own trait for `try_clone`. The `try_clone` method also might return an error which isn't really compatible with `clone()`. Last but not least if `Signals` can be cloned I will have to add the`waker.wake()` part back to `SignalDelivery::flush`. This part of the code really wasn't obvious to me when I saw it first.

So this would make things more complicated again. For which use case do you think cloning `Signals` / `SignalDelivery` will be required?